File: Lab 2/ShpaginApp/Models/DTOs/Book.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;
using ShpaginApp.Models.Validators;

namespace ShpaginApp.Models.DTOs
{
  public record BookItemResponse(
    Guid Id,
    string Name,
    IEnumerable<AuthorItemResponse> Authors
  );

  public record AgeRatingResponse(
    RussianAgeRatingEnum Value,
    string Label
  );

  public record BookResponse
  {
    public Guid Id { get; init; }

    public required string Name { get; init; }

    public int? Year { get; init; }

    public AgeRatingResponse? AgeRating { get; init; }

    public string? Publisher { get; init; }

    public string? Annotation { get; init; }

    public required IEnumerable<AuthorItemResponse> Authors { get; init; }

    public IEnumerable<TagResponse>? Tags { get; init; }

    public DateTime? CreatedAt { get; init; }

    public DateTime? UpdatedAt { get; init; }

    public BookStatistics? Statistics { get; init; }

  };

  public record BookStatistics(
    double AverageRating,
    int NumRatings,
    int InFavorites,
    int TotalReaders,
    int WantToRead,
    int ReadingNow,
    int Completed
  );

  public record CreateBookRequest
  {
    [Required, MaxLength(255)]
    public string? Name { get; init; } = null!;

    [Required, YearRange(minYear: 1000)]
    public int Year { get; init; }

    [Required, Enum(typeof(RussianAgeRatingEnum))]
    public RussianAgeRatingEnum? AgeRating { get; init; } = null!;

    [Required, MinLength(1, ErrorMessage = "At least one author id required")]
    public IEnumerable<Guid> AuthorIds { get; init; } = null!;

    [MinLength(1, ErrorMessage = "At least one tag id required")]
    public IEnumerable<Guid>? TagIds { get; init; }

    [MaxLength(255)]
    public string? Publisher { get; init; }

    [MaxLength(1023)]
    public string? Annotation { get; init; }
  };

  public record UpdateBookPutRequest
  {
    [Required, MaxLength(255)]
    public string Name { get; init; } = null!;

    [Required, YearRange(minYear: 1000)]
    public int? Year { get; init; } = null!;

    [Required, Enum(typeof(RussianAgeRatingEnum))]
    public RussianAgeRatingEnum AgeRating { get; init; }

    [Required, MinLength(1, ErrorMessage = "At least one author id required")]
    public IEnumerable<Guid> AuthorIds { get; init; } = null!;

    [Required]
    public IEnumerable<Guid> TagIds { get; init; } = null!;

    [JsonRequired, MaxLength(255)]
    public string? Publisher { get; init; }

    [JsonRequired, MaxLength(1023)]
    public string? Annotation { get; init; }
  };

  public record UpdateBookPatchRequest
  {
    [MaxLength(255)]
    public string? Name { get; init; }

    [YearRange(minYear: 1000)]
    public int? Year { get; init; }

    [Enum(typeof(RussianAgeRatingEnum))]
    public RussianAgeRatingEnum? AgeRating { get; init; }

    [MinLength(1, ErrorMessage = "At least one author id required")]
    public IEnumerable<Guid>? AuthorIds { get; init; }

    public IEnumerable<Guid>? TagIds { get; init; }

    [MaxLength(255)]
    public string? Publisher { get; init; }

    [MaxLength(1023)]
    public string? Annotation { get; init; }
  };

  public record BookSearchRequest
  {
    [MaxLength(255)]
    public string? Query { get; init; }

    [MinLength(1, ErrorMessage = "At least one author id required")]
    public IEnumerable<Guid>? AuthorIds { get; init; }

    [MinLength(1, ErrorMessage = "At least one tag id required")]
    public IEnumerable<Guid>? TagIds { get; init; }

    [Range(1, 1000)]
    public int? Count { get; init; }
  };
}

File: Lab 3/shpagin-app/src/types/api/book.ts
import { Author } from "@/types/api/author";
import { Tag } from "@/types/api/tag";

export type AgeRating =
  | { value: "ZERO"; label: "0+" }
  | { value: "SIX"; label: "6+" }
  | { value: "TWELVE"; label: "12+" }
  | { value: "SIXTEEN"; label: "16+" }
  | { value: "EIGHTEEN"; label: "18+" };

export const ageRatingValues = [
  "ZERO",
  "SIX",
  "TWELVE",
  "SIXTEEN",
  "EIGHTEEN",
] as const;

export const ageRatingList = [
  { value: "ZERO", label: "0+" },
  { value: "SIX", label: "6+" },
  { value: "TWELVE", label: "12+" },
  { value: "SIXTEEN", label: "16+" },
  { value: "EIGHTEEN", label: "18+" },
] as const;

export interface BookStatistics {
  average_rating: number;
  num_ratings: number;
  in_favorites: number;
  total_readers: number;
  want_to_read: number;
  reading_now: number;
  completed: number;
}

export interface Book {
  id: string;
  name: string;
  year: number | null;
  age_rating: AgeRating | null;
  publisher: string | null;
  annotation: string | null;
  authors: Author[];
  tags: Tag[] | null;
  created_at: string | null;
  updated_at: string | null;
  statistics: BookStatistics | null;
}

export interface BookItem {
  id: string;
  name: string;
  authors: Author[];
}

export interface PaginatedResult<T> {
  page: number;
  page_size: number;
  total_count: number;
  total_pages: number;
  has_previous_page: boolean;
  has_next_page: boolean;
  items: T[];
}
